incubator/hnc/pkg/validators/hierarchy.go

```diff
@@ -172,97 +176,110 @@ func (v *Hierarchy) checkParent(ns, curParent, newParent *forest.Namespace) admi
 
 // authzReq represents a request for authorization
 type authzReq struct {
-	nnm    string // the namespace the user needs to be authorized to modify
-	reason string // the reason we're checking it (for logs and error messages)
+	nnm     string // the namespace the user needs to be authorized to modify
+	reason  string // the reason we're checking it (for logs and error messages)
+	missing bool   // if true, the authz will fail due to this namespace not existing yet (see below)
 }
 
 // needAuthzOn returns the namespaces that the user must be authorized to update in order to make
 // this change. It must be called while the forest lock is held.
 //
+// The general rule is that the user must be an admin of the most recent common ancestor of the old
+// and new parent, if they're both in the same tree. If they're in *different* trees, the user must
+// be an admin of the root of the old tree, and of the new parent. See the user guide or design doc
+// for the rationale for this choice.
+//
+// While this is fairly simple in theory, missing parents make this a bit more complicated.
+//
+// If this webhook is working correctly, a namespace can never be deliberately assigned to a parent
+// that doesn't exist (in Gitops flows, this means that the client is expected to create all
+// namespaces before syncing HierarchyConfiguration objects, or at least be able to keep retrying
+// until after all namespaces have been created). Therefore, there are only three cases where an
+// ancestor might be missing:
+//
+// 1. HNC is just starting up and the parent hasn't been synced yet, so we can't determine the tree
+// root. In these cases, we should just reject the request and ask the user to try again.
+// 2. The parent has been deleted, and the namespace is orphaned. In this case, the namespace will
+// have a CritParentMissing condition, and we should treat the namespace as though it's already a
```

Comment:
In the case 1, when the parent is not synced yet, the namespace will have `CritParentMissing` too, right? How are case 1 and 2 handled differently? Could you please point it to me where case 1 is denied in the code?

Reply:
Good catch! You're right, for some reason I was thinking that the condition wouldn't be set, but of course that's not true.

What do you think we should do? I can't think of a good way to tell the difference in our current framework. Do you think it's fine to simply not support this case?

Reply:
Do you mean just deny both case 1 and 2? I couldn't think of a good way to handle case 2 either. Denying them (both case 1 and 2) sgtm.

Reply:
I'd meant the opposite - if there's a missing parent, let an admin resolve the problem. If we go with "always deny," how do you think an admin should fix it?

Reply:
Ah ok, sgtm. Then the only concern is moving a namespace when the parent is not synced, right? Just confirming.

Reply:
This logic's been fully revamped as discussed, PTAL